state/prunning_service.go

```diff
@@ -0,0 +1,228 @@
+package state
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/libs/service"
+)
+
+type PruningRequester int64
+
+const (
+	AppRequester                 PruningRequester = 0
+	DataCompanionRequester       PruningRequester = 1
+	ABCIResRetainHeightRequester PruningRequester = 2
+)
+
+var (
+	AppRetainHeightKey           = []byte("AppRetainHeightKey")
+	DataCompanionRetainHeightKey = []byte("DCRetainHeightKey")
+	ABCIResultsRetainHeightKey   = []byte("ABCIResRetainHeightKey")
+)
+
+type Pruner struct {
+	service.BaseService
+	logger log.Logger
+	// DB to which we save the retain heights
+	bs BlockStore
+	// State store to prune state from
+	stateStore    Store
+	pruningTicker *time.Ticker
+}
+
+type RetainHeightInfo struct {
+	Height    int64
+	Requester PruningRequester
+}
+
+func NewPruner(stateStore Store, bs BlockStore, logger log.Logger) *Pruner {
+	p := &Pruner{
+		bs:            bs,
+		stateStore:    stateStore,
+		logger:        logger,
+		pruningTicker: time.NewTicker(time.Second * 10),
```

Comment:
Frequency should be customizable.

Reply:
Yeah it should just be a parameter supplied to the `NewPruner` constructor. You could also use the options pattern (like in the [gRPC server constructor](https://github.com/cometbft/cometbft/blob/902631a5f58acf0888c2ae0a98a63ff55f4a101f/rpc/grpc/server/server.go#L77) and elsewhere in the code) for all optional parameters past the `logger`.

That way, usage like the following would be possible:

```golang
// Create a pruner with the default pruning interval.
pruner := state.NewPruner(stateStore, blockStore, logger)

// Create a pruner, overriding the default pruning interval.
pruner := state.NewPruner(stateStore, blockStore, logger, state.PruneInterval(60 * time.Second))
```

Alternatively you could make it a compulsory parameter, but it seems like it'd be useful to make it optional in the constructor.